Framework/Algorithms/src/FitPeaks.cpp

```diff
@@ -0,0 +1,2232 @@
+ * @param dataws :: matrix workspace to fit with
+ * @param wsindex ::  workspace index of the spectrum in matrix workspace
+ * @param vec_xmin :: minimin values of domains
+ * @param vec_xmax :: maximim values of domains
+ * @return
+ */
+double FitPeaks::fitFunctionMD(API::IFunction_sptr fit_function,
+                               API::MatrixWorkspace_sptr dataws, size_t wsindex,
+                               std::vector<double> &vec_xmin,
+                               std::vector<double> &vec_xmax) {
+  // Validate
+  if (vec_xmin.size() != vec_xmax.size())
+    throw runtime_error("Sizes of xmin and xmax (vectors) are not equal. ");
+
+  // Note: after testing it is found that multi-domain Fit cannot be reused
+  API::IAlgorithm_sptr fit;
+  try {
+    fit = createChildAlgorithm("Fit", -1, -1, false);
+  } catch (Exception::NotFoundError &) {
+    std::stringstream errss;
+    errss << "The FitPeak algorithm requires the CurveFitting library";
+    throw std::runtime_error(errss.str());
+  }
+  // set up background fit instance
+  fit->setProperty("Minimizer", m_minimizer);
+  fit->setProperty("CostFunction", m_costFunction);
+  fit->setProperty("CalcErrors", true);
+
+  // This use multi-domain; but does not know how to set up IFunction_sptr
+  // fitfunc,
+  boost::shared_ptr<MultiDomainFunction> md_function =
+      boost::make_shared<MultiDomainFunction>();
+
+  // Set function first
+  md_function->addFunction(fit_function);
+
+  //  set domain for function with index 0 covering both sides
+  md_function->clearDomainIndices();
+  std::vector<size_t> ii(2);
+  ii[0] = 0;
+  ii[1] = 1;
+  md_function->setDomainIndices(0, ii);
+
+  // Set the properties
+  fit->setProperty("Function",
+                   boost::dynamic_pointer_cast<IFunction>(md_function));
+  fit->setProperty("InputWorkspace", dataws);
+  fit->setProperty("WorkspaceIndex", static_cast<int>(wsindex));
+  fit->setProperty("StartX", vec_xmin[0]);
+  fit->setProperty("EndX", vec_xmax[0]);
+  fit->setProperty("InputWorkspace_1", dataws);
+  fit->setProperty("WorkspaceIndex_1", static_cast<int>(wsindex));
+  fit->setProperty("StartX_1", vec_xmin[1]);
+  fit->setProperty("EndX_1", vec_xmax[1]);
+  fit->setProperty("MaxIterations", 50);
+
+  // Execute
+  fit->execute();
+  if (!fit->isExecuted()) {
+    throw runtime_error("Fit is not executed on multi-domain function/data. ");
+  }
+
+  // Retrieve result
+  std::string fitStatus = fit->getProperty("OutputStatus");
+
+  double chi2 = DBL_MAX;
+  if (fitStatus == "success") {
+    chi2 = fit->getProperty("OutputChi2overDoF");
+  }
+
+  return chi2;
+}
+
+//----------------------------------------------------------------------------------------------
+/** Fit peak with high background
+ * @brief FitPeaks::FitFunctionHighBackground
+ * @param fit
+ * @param fit_window
+ * @param ws_index
+ * @param expected_peak_center
+ * @param observe_peak_width
+ * @param peakfunction
+ * @param bkgdfunc
+ * @param 
```

Comment:
Why copy? The entire loop is not necessary if you use `DataObjects::create`, since it copies X if the number of histograms is unchanged. See https://github.com/mantidproject/documents/blob/master/Presentations/DevMeetings/2018-01/workspace-factory.pdf.